ssl/ssl_lib.c

```diff
@@ -4343,6 +4343,80 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
     return ctx->ctlog_store;
 }
 
+void
+SSL_CTX_set_early_cb(SSL_CTX *c, ssl_early_cb_fn cb, void *arg)
+{
+    c->early_cb = cb;
+    c->early_cb_arg = arg;
+}
+
+int
+SSL_EARLY_isv2(SSL *s, SSL_EARLY *ectx)
+{
+    return ectx->isv2;
+}
+
+unsigned int
+SSL_EARLY_get0_version(SSL *s, SSL_EARLY *ectx)
+{
+    return ectx->legacy_version;
+}
+
+size_t
+SSL_EARLY_get0_random(SSL *s, SSL_EARLY *ectx, const unsigned char **out)
+{
+    if (out != NULL)
+        *out = ectx->random;
+    return ectx->isv2 ? 0 : SSL3_RANDOM_SIZE;
+}
+
+size_t
+SSL_EARLY_get0_session_id(SSL *s, SSL_EARLY *ectx, const unsigned char **out)
+{
+    if (out != NULL)
+        *out = ectx->session_id;
+    return ectx->session_id_len;
+}
+
+size_t
+SSL_EARLY_get0_ciphers(SSL *s, SSL_EARLY *ectx, const unsigned char **out)
+{
+    if (out != NULL)
+        *out = PACKET_data(&ectx->ciphersuites);
+    return PACKET_remaining(&ectx->ciphersuites);
+}
+
+size_t
+SSL_EARLY_get0_compression_methods(SSL *s, SSL_EARLY *ectx,
+                                   const unsigned char **out)
+{
+    if (out != NULL)
+        *out = ectx->compressions;
+    return ectx->compressions_len;
+}
+
+int
+SSL_EARLY_get0_ext(SSL *s, SSL_EARLY *ectx, unsigned int type,
+                   const unsigned char **out, size_t *outlen)
+{
+    size_t i;
+    RAW_EXTENSION *r;
+
+    /* XXX potential accidental quadratic behavior */
+    for (i = 0; i < ectx->pre_proc_exts_len; ++i) {
```

Comment:
I'll remove the XXX comment, given this and davidben's comment.

> That said and index -> type lookup might not be a bad idea....or maybe expose the index????

Do you mean type->index lookup?  I had thought about doing that at some point, but demurred given that (1) there wasn't one already, possibly intentionally, and (2) I didn't want to think about how it would interact with custom extensions right then.